architecting-cloud-native-dotnet-apps-for-azure/cloud-native-devops.md

```diff
@@ -5,8 +5,246 @@ ms.date: 06/30/2019
+
+Each of these components provides some advantages for Cloud Native Applications but the three most useful are the source control, boards and pipelines.  
+
+## Source Control
+
+Organize the code for a Cloud Native Application can be somewhat challenging. Instead of a single giant application the Cloud Native Applications tend to be made up of a web of smaller applications which talk with one another. As with all things in computing the best arrangement of code remains an open question. There are examples of successful applications using a variety of layouts but two variants seem to have the most popularity. 
+
+Before getting down into the actual source control itself it is probably worth deciding on how many projects are appropriate. Within a single project is support for multiple repositories, and build pipelines. Board are a little more complicated but there too the tasks can easily be assigned to multiple teams within a single project. It is certainly possible to support hundreds, even thousands of developers out of a single Azure DevOps project. Doing so is likely the best approach as it provides a single place for all developer to work out of and reduces the confusion of finding that one application when developers are unsure of even which project in which it resides. 
+
+With the single project argument settled the question moves to how to split up the code for the microservices. 
+
+![](media/single-repository-vs-multiple.png)
+
+### Repository per Microservice
+
+At first glance this seems like the most logical approach to splitting up the source code for microservices. Each repository can contain the code needed to build the one microservice. The advantages to this approach are readily visible: 
+
+1. Instructions for building and maintaining the application can be added to a README file at the root of each repository. When flipping through the repositories it will be easy to find these instruction reducing spin up time for developers. 
+2. Every service is located in a logical place, easily found by knowing the name of the service. 
+3. Builds can easily be set up such that they're only triggered when a change is made to the owning repository.
+4. The number of changes coming into a repository is limited to the small number of developers working on the project. 
+5. Security is easy to set up by restricting the repositories to which developers have read and write permissions.
+6. Repository level settings can be changed by the owning team with a minimum of discussion with others
+
+One of the key ideas behind microservices is that services should be siloed and separated from each other. This idea is reflected by the isolation of microservice data to a database separate and autonomous from the rest of the services. It makes a great deal of sense to carry this idea all the way through to the source code.
```

Comment:
May be worth mentioning that in DDD terms, each microservice should act as its own bounded context.